src/org/jitsi/impl/neomedia/rtp/RTPEncodingImpl.java

```diff
@@ -0,0 +1,469 @@
+                if (enc.requires(idx))
+                {
+                    requires = true;
+                    break;
+                }
+            }
+        }
+
+        return requires;
+    }
+
+    /**
+     * {@inheritDoc}
+     */
+    @Override
+    public boolean matches(byte[] buf, int off, int len)
+    {
+        long ssrc = RawPacket.getSSRCAsLong(buf, off, len);
+
+        if (primarySSRC != ssrc && rtxSSRC != ssrc)
+        {
+            return false;
+        }
+
+        if (temporalId == -1)
+        {
+            return true;
+        }
+
+        int tid = track.getMediaStreamTrackReceiver()
+            .getStream().getTemporalID(buf, off, len);
+
+        return tid == temporalId;
+    }
+
+    /**
+     * Gets a boolean flag that indicates whether or not this instance is
+     * streaming or if it's suspended.
+     *
+     * @param active true if this {@link RTPEncodingImpl} is active, otherwise
+     * false.
+     */
+    void setActive(boolean active)
+    {
+        this.active = active;
+    }
+
+    /**
+     *
+     * @param pkt
+     * @param nowMs
+     *
+     * @return the {@link FrameDesc} that was updated, otherwise null.
+     */
+    FrameDesc update(RawPacket pkt, long nowMs)
+    {
+        // Update rate stats.
+        rateStatistics.update(pkt.getLength(), nowMs);
+
+        long ts = pkt.getTimestamp();
+        FrameDesc frame = frames.get(ts);
+
+        if (frame == null)
+        {
+            synchronized (frames)
+            {
+                frames.put(ts, frame = new FrameDesc(this, ts));
+            }
+
+            // We measure the stable bitrate on every new frame.
+            lastStableBitrateBps = getBitrateBps(nowMs);
+        }
+
+        // Update the frame description.
+        boolean frameChanged = frame.update(pkt);
+        if (frameChanged)
+        {
+            // Frame boundaries heuristics.
+
+            // Find the closest next frame.
+            Map.Entry<Long, FrameDesc> ceilingEntry
+                = frames.ceilingEntry((ts + 1) & 0xFFFFFFFFL);
+
+            if (ceilingEntry != null)
+            {
+                FrameDesc ceilingFrame = ceilingEntry.getValue();
+
+                // Make sure the frames are close enough, both in time and in
+                // sequence numbers.
+                long tsDiff = (ceilingFrame.getTimestamp() - ts) & 0xFFFFFFFFL;
+
+                if (tsDiff < (1L<<31)
+                    && (ceilingFrame.getStart() != -1 || frame.getEnd() != -1))
+                {
+                    int minSeen = ceilingFrame.getMinSeen();
+                    int maxSeen = frame.getMaxSeen();
+                    int snDiff = (minSeen - maxSeen) & 0xFFFF;
+
+                    // We can deal with distances that are less than 3.
+                    if (snDiff < 3)
+                    {
+                        frame.setEnd((maxSeen + 1) & 0xFFFF);
+                        ceilingFrame.setStart((minSeen - 1) & 0xFFFF);
```

Comment:
Why do we unconditionally bump the frame's end here? It could be that we've already seen the end, and frame.getEnd() == frame.maxSeen().
Same for the start of the ceiling frame. In fact, with diff <= 2, at least one of the bumps should not happen. Or am I missing something?